File: backend/app/deezer.py
"""Music catalog client.

Despite the module name, this calls **Deezer's** public API rather than
Spotify. The original Spotify Client-Credentials path was abandoned in late
2024 after Spotify silently stripped `preview_url`, `popularity`, and `genres`
from non-extended-quota apps — making the entire game premise (play 30-second
clips, gate by popularity) impossible.

Deezer offers what we need with no auth required:
  - `/search?q=...`         tracks with `preview` (30s MP3) and `rank` (popularity-ish)
  - `/search/artist?q=...`  artists with id, name, picture URLs
  - `/track/{id}`           full track including preview

The module name is unchanged to minimise churn elsewhere — column names and
fields like `spotify_track_id` now hold Deezer integer IDs (as strings).
"""

import asyncio
import logging
import random
from typing import Any

import httpx

from app.matching import normalize_artist, normalize_title

logger = logging.getLogger("slomp.music")

_API = "https://api.deezer.com"


async def _api_get(path: str, params: dict | None = None) -> Any:
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.get(f"{_API}{path}", params=params)
        if resp.status_code >= 400:
            logger.error(
                "deezer GET %s -> %s body=%s",
                resp.request.url,
                resp.status_code,
                resp.text[:500],
            )
        resp.raise_for_status()
        data = resp.json()
        # Deezer signals errors with HTTP 200 + {"error": {...}} body
        if isinstance(data, dict) and data.get("error"):
            err = data["error"]
            logger.error("deezer api error %s: %s", err.get("code"), err.get("message"))
            raise httpx.HTTPStatusError(
                f"Deezer error {err.get('code')}: {err.get('message')}",
                request=resp.request,
                response=resp,
            )
        return data


async def search_tracks(query: str, limit: int = 10) -> list[dict]:
    data = await _api_get("/search", {"q": query, "limit": limit})
    return data.get("data", [])


async def search_tracks_for_artists(
    query: str, artist_names: list[str], limit_per_artist: int = 25
) -> list[dict]:
    """Run one Deezer search per required artist with the artist name
    appended free-text to the query (`q=<query> <name>`), then merge unique
    tracks. We don't use Deezer's `artist:"<name>"` advanced filter because
    its name match is token-based and fuzzy — `artist:"Daft Punk"` happily
    returns tracks by "Pan Da Punk". Free-text concatenation lets Deezer's
    own relevance rank the real artist first; the caller still strips
    look-alikes by ID via `matches_rules`."""
    if not artist_names:
        return []

    async def _one(name: str) -> list[dict]:
        scoped = f"{query} {name}".strip()
        if not scoped:
            return []
        try:
            return await search_tracks(scoped, limit=limit_per_artist)
        except httpx.HTTPStatusError as e:
            logger.warning("artist-scoped search failed for %r: %s", name, e)
            return []

    batches = await asyncio.gather(*(_one(n) for n in artist_names))
    merged: dict[Any, dict] = {}
    for batch in batches:
        for t in batch:
            tid = t.get("id")
            if tid is not None and tid not in merged:
                merged[tid] = t
    return list(merged.values())


def relevance_score(track: dict, query: str) -> float:
    """Score a Deezer track against the user's query for our own ranking.
    Deezer's relevance collapses once you add an `artist:` filter, so the
    merged-search path needs to re-rank with title/token overlap and use
    `rank` only as a tiebreaker."""
    q = (query or "").strip().lower()
    title = (track.get("title") or "").lower()
    artist_name = ((track.get("artist") or {}).get("name") or "").lower()
    score = 0.0
    if q:
        if title == q:
            score += 100.0
        elif title.startswith(q):
            score += 60.0
        elif q in title:
            score += 35.0
        q_tokens = set(q.split())
        if q_tokens:
            t_tokens = set(title.split())
            score += 25.0 * len(q_tokens & t_tokens) / len(q_tokens)
        if q in artist_name:
            score += 5.0
    rank = track.get("rank") or 0
    score += rank / 1_000_000.0
    return score


async def search_artists(query: str, limit: int = 10) -> list[dict]:
    data = await _api_get("/search/artist", {"q": query, "limit": limit})
    return data.get("data", [])


async def get_track(track_id: str) -> dict:
    return await _api_get(f"/track/{track_id}")


# Genre IDs used to assemble a diverse pool of "popular" tracks for the
# random gamemode. 0 = the global all-genres chart; the rest are major
# Deezer genre buckets. Pulling from several at once gets us ~600 unique
# tracks before filtering, even though each individual chart caps at 100.
_RANDOM_GENRE_IDS = (0, 132, 116, 152, 113, 165, 84, 173)


async def fetch_random_tracks(min_popularity: int, count: int) -> list[dict]:
    """Pull a pool of popular tracks from Deezer charts (multiple genres),
    then for each candidate find the version that `/search` surfaces and
    keep only ones whose search-discoverable rank passes `min_popularity`.

    Deezer often has several track IDs for the same recording with
    different ranks (e.g. an album release vs a deluxe-edition release).
    The chart endpoint exposes IDs that don't always appear in search
    results, so without this re-search a random pick can be a song
    classic-mode players cannot submit themselves — an inconsistency a
    guesser hits when their search box doesn't return the song they're
    hearing. Returning the search version also makes the stored
    `popularity` match what players see."""
    rules = {"min_popularity": min_popularity, "required_artists": []}

    async def _chart(gid: int) -> list[dict]:
        try:
            data = await _api_get(f"/chart/{gid}/tracks", {"limit": 100})
            return data.get("data", [])
        except Exception:
            logger.warning("chart fetch failed for genre=%s", gid)
            return []

    batches = await asyncio.gather(*(_chart(g) for g in _RANDOM_GENRE_IDS))
    seen_chart: set = set()
    candidates: list[dict] = []
    for batch in batches:
        for t in batch:
            tid = t.get("id")
            if tid is None or tid in seen_chart:
                continue
            seen_chart.add(tid)
            if not t.get("preview"):
                continue
            candidates.append(t)
    random.shuffle(candidates)

    async def _resolve(t: dict) -> dict | None:
        title = t.get("title") or ""
        artist_name = (t.get("artist") or {}).get("name") or ""
        if not title or not artist_name:
            return None
        norm_title = normalize_title(title)
        norm_artist = normalize_artist(artist_name)
        try:
            results = await search_tracks(f"{title} {artist_name}".strip(), limit=5)
        except Exception:
            return None
        for r in results:
            if normalize_title(r.get("title") or "") != norm_title:
                continue
            r_artist = (r.get("artist") or {}).get("name") or ""
            if normalize_artist(r_artist) != norm_artist:
                continue
            ok, _ = matches_rules(r, rules)
            if ok:
                return r
        return None

    accepted: list[dict] = []
    seen_accepted: set = set()
    chunk = 25
    for i in range(0, len(candidates), chunk):
        if len(accepted) >= count:
            break
        resolved = await asyncio.gather(
            *(_resolve(t) for t in candidates[i : i + chunk])
        )
        for r in resolved:
            if r is None:
                continue
            tid = r.get("id")
            if tid is None or tid in seen_accepted:
                continue
            seen_accepted.add(tid)
            accepted.append(r)
            if len(accepted) >= count:
                break
    return accepted[:count]


async def get_artist(artist_id: str) -> dict:
    return await _api_get(f"/artist/{artist_id}")


def matches_rules(track: dict, rules: dict) -> tuple[bool, str | None]:
    """Validate a track against room rules. (popularity, required_artists,
    preview availability — genres are unsupported since Deezer only exposes
    them at the album level via numeric IDs.)"""
    if not track.get("preview"):
        return False, "track has no preview audio"

    # Deezer's `rank` is roughly 0..1_000_000+ for big hits. Map our
    # min_popularity (0..100) onto that linearly.
    rank = track.get("rank", 0) or 0
    min_popularity = rules.get("min_popularity", 0)
    min_rank = min_popularity * 10_000
    if rank < min_rank:
        return False, "track popularity below minimum"

    required = set(rules.get("required_artists") or [])
    if required:
        artist_ids: set[str] = set()
        primary = track.get("artist", {})
        if primary.get("id") is not None:
            artist_ids.add(str(primary["id"]))
        for c in track.get("contributors", []) or []:
            if c.get("id") is not None:
                artist_ids.add(str(c["id"]))
        if not (artist_ids & required):
            return False, "not by a required artist"

    return True, None


def serialize_candidate(track: dict) -> dict:
    """Shape a Deezer track to our SongCandidate schema."""
    artist = track.get("artist") or {}
    album = track.get("album") or {}
    image = album.get("cover_medium") or album.get("cover") or album.get("cover_big")
    rank = track.get("rank") or 0
    duration_seconds = track.get("duration") or 0
    return {
        "spotify_track_id": str(track["id"]),  # field-name vestigial; holds Deezer id
        "title": track.get("title", ""),
        "artist": artist.get("name", ""),
        "album": album.get("title"),
        "preview_url": track.get("preview"),
        "album_image_url": image,
        "duration_ms": duration_seconds * 1000,
        "popularity": min(100, rank // 10_000),
    }

File: frontend/src/lib/deezer.ts
// Deezer public-API client, called straight from the browser.
//
// Previously the backend proxied every search; with 7+ players in a room
// hammering the search box that bucket emptied fast and Deezer started
// 4xx-ing the whole room. Each client now spends from its own per-IP
// allowance instead.
//
// Deezer's API does not send `Access-Control-Allow-Origin`, so a normal
// `fetch` is blocked by CORS. They support `?output=jsonp&callback=…` —
// inject a <script> tag, the response invokes our callback. The page CSP
// has `https://api.deezer.com` whitelisted in `script-src` for this.

import type { ArtistSummary, RoomSettings, SongCandidate } from "./types";

const API = "https://api.deezer.com";

// Sliding-window limiter sized to roughly match Deezer's documented
// per-IP cap (50 calls / 5 s). One *user search* with required_artists
// can fan out into N+1 calls, so we count individual API calls, not
// user-visible searches.
const RATE_LIMIT_MAX = 50;
const RATE_LIMIT_WINDOW_MS = 5_000;
const recentCalls: number[] = [];

export class RateLimitError extends Error {
  constructor() {
    super("Slow down — 50 searches per 5 s");
    this.name = "RateLimitError";
  }
}

function reserveSlot() {
  const now = Date.now();
  while (recentCalls.length && recentCalls[0] < now - RATE_LIMIT_WINDOW_MS) {
    recentCalls.shift();
  }
  if (recentCalls.length >= RATE_LIMIT_MAX) throw new RateLimitError();
  recentCalls.push(now);
}

let cbCounter = 0;

interface DeezerError {
  error?: { code: number | string; message: string; type?: string };
}

const JSONP_TIMEOUT_MS = 8_000;

function jsonp<T>(
  path: string,
  params: Record<string, string | number> = {},
): Promise<T> {
  reserveSlot();
  return new Promise<T>((resolve, reject) => {
    const cbName = `__deezer_cb_${++cbCounter}`;
    const usp = new URLSearchParams();
    for (const [k, v] of Object.entries(params)) usp.set(k, String(v));
    usp.set("output", "jsonp");
    usp.set("callback", cbName);

    const script = document.createElement("script");
    let settled = false;
    let timer: ReturnType<typeof setTimeout> | null = null;
    const cleanup = () => {
      if (timer != null) clearTimeout(timer);
      delete (window as unknown as Record<string, unknown>)[cbName];
      script.remove();
    };

    (window as unknown as Record<string, (data: T & DeezerError) => void>)[
      cbName
    ] = (data) => {
      if (settled) return;
      settled = true;
      cleanup();
      if (data && typeof data === "object" && "error" in data && data.error) {
        reject(new Error(`Deezer ${data.error.code}: ${data.error.message}`));
        return;
      }
      resolve(data as T);
    };

    script.onerror = () => {
      if (settled) return;
      settled = true;
      cleanup();
      reject(new Error("Deezer JSONP request failed"));
    };

    // Without a hard timeout the script tag will hang forever if Deezer
    // never responds (e.g. extension blocking the request) — and the
    // caller's UI sits on a "searching…" spinner indefinitely.
    timer = setTimeout(() => {
      if (settled) return;
      settled = true;
      cleanup();
      reject(new Error("Deezer request timed out"));
    }, JSONP_TIMEOUT_MS);

    script.src = `${API}${path}?${usp.toString()}`;
    document.head.appendChild(script);
  });
}

interface DeezerTrack {
  id: number;
  title: string;
  rank?: number;
  preview?: string | null;
  duration?: number;
  artist?: { id: number; name: string };
  album?: {
    title?: string;
    cover?: string;
    cover_medium?: string;
    cover_big?: string;
  };
  contributors?: { id: number; name: string }[];
}

interface DeezerArtist {
  id: number;
  name: string;
  picture?: string;
  picture_medium?: string;
}

interface DeezerSearchResponse<T> {
  data: T[];
}

async function searchTracksRaw(
  q: string,
  limit: number,
): Promise<DeezerTrack[]> {
  const data = await jsonp<DeezerSearchResponse<DeezerTrack>>("/search", {
    q,
    limit,
  });
  return data.data ?? [];
}

async function searchArtistsRaw(
  q: string,
  limit: number,
): Promise<DeezerArtist[]> {
  const data = await jsonp<DeezerSearchResponse<DeezerArtist>>(
    "/search/artist",
    { q, limit },
  );
  return data.data ?? [];
}

async function getArtistRaw(id: string): Promise<DeezerArtist> {
  return jsonp<DeezerArtist>(`/artist/${id}`);
}

// Re-rank merged artist-scoped batches: Deezer's relevance gets noisy once
// you concatenate the artist name onto the query, so we score title/token
// overlap ourselves and use rank only as a tiebreaker.
function relevanceScore(track: DeezerTrack, query: string): number {
  const q = query.trim().toLowerCase();
  const title = (track.title ?? "").toLowerCase();
  const artistName = (track.artist?.name ?? "").toLowerCase();
  let score = 0;
  if (q) {
    if (title === q) score += 100;
    else if (title.startsWith(q)) score += 60;
    else if (title.includes(q)) score += 35;

    const qTokens = new Set(q.split(/\s+/).filter(Boolean));
    if (qTokens.size > 0) {
      const tTokens = new Set(title.split(/\s+/).filter(Boolean));
      let overlap = 0;
      for (const t of qTokens) if (tTokens.has(t)) overlap += 1;
      score += (25 * overlap) / qTokens.size;
    }
    if (artistName.includes(q)) score += 5;
  }
  score += (track.rank ?? 0) / 1_000_000;
  return score;
}

function matchesRules(track: DeezerTrack, settings: RoomSettings): boolean {
  if (!track.preview) return false;

  // `rank` is roughly 0..1_000_000+; map min_popularity (0..100) onto it
  // linearly. Backend uses the same factor.
  const rank = track.rank ?? 0;
  if (rank < settings.min_popularity * 10_000) return false;

  const required = settings.required_artists ?? [];
  if (required.length > 0) {
    const ids = new Set<string>();
    if (track.artist?.id != null) ids.add(String(track.artist.id));
    for (const c of track.contributors ?? []) {
      if (c.id != null) ids.add(String(c.id));
    }
    if (!required.some((r) => ids.has(r))) return false;
  }
  return true;
}

function serializeCandidate(track: DeezerTrack): SongCandidate {
  const album = track.album ?? {};
  const image = album.cover_medium ?? album.cover ?? album.cover_big ?? null;
  const rank = track.rank ?? 0;
  const durationSec = track.duration ?? 0;
  return {
    track_id: String(track.id),
    title: track.title ?? "",
    artist: track.artist?.name ?? "",
    album: album.title ?? null,
    preview_url: track.preview ?? null,
    album_image_url: image,
    duration_ms: durationSec * 1000,
    popularity: Math.min(100, Math.floor(rank / 10_000)),
  };
}

function serializeArtist(a: DeezerArtist): ArtistSummary {
  return {
    id: String(a.id),
    name: a.name,
    image_url: a.picture_medium ?? a.picture ?? null,
    genres: [],
    popularity: null,
  };
}

export async function searchSongs(
  query: string,
  settings: RoomSettings | null,
  limit = 10,
): Promise<SongCandidate[]> {
  const required = settings?.required_artists ?? [];

  let tracks: DeezerTrack[];
  if (required.length > 0) {
    // Resolve required IDs → names, then run one search per artist with
    // the name appended free-text (`q=<query> <name>`). We don't use
    // Deezer's `artist:"<name>"` advanced filter because its match is
    // token-fuzzy and lets look-alikes through; matchesRules strips
    // covers/tributes by ID afterwards.
    const artistResults = await Promise.allSettled(
      required.map((id) => getArtistRaw(id)),
    );
    const names = artistResults
      .map((r) => (r.status === "fulfilled" ? r.value.name : null))
      .filter((n): n is string => !!n);

    const batches = await Promise.all(
      names.map(async (name) => {
        const scoped = `${query} ${name}`.trim();
        if (!scoped) return [] as DeezerTrack[];
        try {
          return await searchTracksRaw(scoped, 25);
        } catch {
          return [] as DeezerTrack[];
        }
      }),
    );

    const merged = new Map<number, DeezerTrack>();
    for (const batch of batches) {
      for (const t of batch) {
        if (t.id != null && !merged.has(t.id)) merged.set(t.id, t);
      }
    }
    tracks = [...merged.values()].sort(
      (a, b) => relevanceScore(b, query) - relevanceScore(a, query),
    );
  } else {
    tracks = await searchTracksRaw(query, limit);
  }

  const out: SongCandidate[] = [];
  for (const t of tracks) {
    if (settings && !matchesRules(t, settings)) continue;
    out.push(serializeCandidate(t));
    if (out.length >= limit) break;
  }
  return out;
}

export async function searchArtists(
  query: string,
  limit = 10,
): Promise<ArtistSummary[]> {
  const artists = await searchArtistsRaw(query, limit);
  return artists.map(serializeArtist);
}

export async function getArtists(ids: string[]): Promise<ArtistSummary[]> {
  if (ids.length === 0) return [];
  const results = await Promise.allSettled(ids.map((id) => getArtistRaw(id)));
  return results
    .filter(
      (r): r is PromiseFulfilledResult<DeezerArtist> =>
        r.status === "fulfilled" && r.value?.id != null,
    )
    .map((r) => serializeArtist(r.value));
}
